File: app/src/main/java/de/westnordost/streetcomplete/data/osm/download/ElementGeometryCreator.java
package de.westnordost.streetcomplete.data.osm.download;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

import de.westnordost.streetcomplete.data.meta.OsmAreas;
import de.westnordost.streetcomplete.data.osm.ElementGeometry;
import de.westnordost.osmapi.map.data.Element;
import de.westnordost.osmapi.map.data.LatLon;
import de.westnordost.osmapi.map.data.Node;
import de.westnordost.osmapi.map.data.Relation;
import de.westnordost.osmapi.map.data.RelationMember;
import de.westnordost.osmapi.map.data.Way;
import de.westnordost.streetcomplete.util.SphericalEarthMath;

public class ElementGeometryCreator
{
	protected WayGeometrySource data;

	public void setWayGeometryProvider(WayGeometrySource data)
	{
		this.data = data;
	}

	public ElementGeometry create(Node node)
	{
		return new ElementGeometry(node.getPosition());
	}

	public ElementGeometry create(Way way)
	{
		if(data == null) throw new NullPointerException();

		List<LatLon> polyline = data.getNodePositions(way.getId());
		// unable to create geometry
		if(polyline.isEmpty()) return null;
		eliminateDuplicates(polyline);

		List<List<LatLon>> polylines = new ArrayList<>(1);
		polylines.add(polyline);

		ElementGeometry result;
		if(OsmAreas.isArea(way))
		{
			// ElementGeometry considers polygons that are defined clockwise holes, so ensure that
			// it is defined CCW here.
			if(SphericalEarthMath.isRingDefinedClockwise(polyline)) {
				Collections.reverse(polyline);
			}
			result = new ElementGeometry(null, polylines);
		}
		else
		{
			result = new ElementGeometry(polylines, null);
		}
		if(result.center == null) return null;
		return result;
	}

	private void eliminateDuplicates(List<LatLon> polyline)
	{
		Iterator<LatLon> it = polyline.iterator();
		LatLon previous = null;
		while(it.hasNext())
		{
			LatLon line = it.next();
			if(previous == null ||
					line.getLatitude() != previous.getLatitude() ||
					line.getLongitude() != previous.getLongitude())
			{
				previous = line;
			}
			else
			{
				it.remove();
			}
		}
	}

	public ElementGeometry create(Relation relation)
	{
		if(data == null) throw new NullPointerException();

		if(OsmAreas.isArea(relation))
		{
			return createMultipolygonGeometry(relation);
		}
		else
		{
			return createPolylinesGeometry(relation);
		}
	}

	private List<List<LatLon>> getWaysOfRelationWithRole(Relation relation, String role)
	{
		List<List<LatLon>> result = new ArrayList<>();
		for(RelationMember member : relation.getMembers())
		{
			if(member.getType() != Element.Type.WAY) continue;

			long wayId = member.getRef();
			if(role == null || role.equals(member.getRole()))
			{
				List<LatLon> nodePositions = data.getNodePositions(wayId);
				eliminateDuplicates(nodePositions);
				if(nodePositions.size() > 1)
				{
					result.add(nodePositions);
				}
			}
		}
		return result;
	}

	private ElementGeometry createPolylinesGeometry(Relation relation)
	{
		List<List<LatLon>> waysNodePositions = getWaysOfRelationWithRole(relation, null);
		ConnectedWays ways = joinWays(waysNodePositions);

		List<List<LatLon>> polylines = ways.rest;
		polylines.addAll(ways.rings);

		// no valid geometry
		if(polylines.isEmpty()) return null;

		ElementGeometry result = new ElementGeometry(polylines, null);
		if(result.center == null) return null;

		return result;
	}

	private ElementGeometry createMultipolygonGeometry(Relation relation)
	{
		List<List<LatLon>> rings = new ArrayList<>();
		rings.addAll(createNormalizedRingGeometry(relation, "outer", false));
		rings.addAll(createNormalizedRingGeometry(relation, "inner", true));
		// no valid geometry
		if(rings.isEmpty()) return null;

		ElementGeometry result = new ElementGeometry(null, rings);
		if(result.center == null) return null;

		return result;
	}

	private List<List<LatLon>> createNormalizedRingGeometry(Relation relation, String role,
															boolean clockwise)
	{
		List<List<LatLon>> waysNodePositions = getWaysOfRelationWithRole(relation, role);
		List<List<LatLon>> ringGeometry = joinWays(waysNodePositions).rings;
		setOrientation(ringGeometry, clockwise);
		return ringGeometry;
	}

	/** Ensures that all given rings are defined in clockwise/counter-clockwise direction */
	private static void setOrientation(List<List<LatLon>> rings, boolean clockwise)
	{
		for(List<LatLon> ring : rings)
		{
			if(SphericalEarthMath.isRingDefinedClockwise(ring) != clockwise)
			{
				Collections.reverse(ring);
			}
		}
	}

	private static class ConnectedWays
	{
		List<List<LatLon>> rings = new ArrayList<>();
		List<List<LatLon>> rest = new ArrayList<>();
	}

	private static ConnectedWays joinWays(List<List<LatLon>> waysNodePositions)
	{
		NodeWayMap<LatLon> nodeWayMap = new NodeWayMap<>(waysNodePositions);

		ConnectedWays result = new ConnectedWays();

		List<LatLon> currentWay = new ArrayList<>();

		while(nodeWayMap.hasNextNode())
		{
			LatLon node;
			if(currentWay.isEmpty())
			{
				node = nodeWayMap.getNextNode();
			}
			else
			{
				node = currentWay.get(currentWay.size()-1);
			}

			List<List<LatLon>> waysAtNode = nodeWayMap.getWaysAtNode(node);
			if(waysAtNode == null)
			{
				result.rest.add(currentWay);
				currentWay = new ArrayList<>();
			}
			else
			{
				List<LatLon> way = waysAtNode.get(0);

				addTo(way, currentWay);
				nodeWayMap.removeWay(way);

				// finish ring and start new one
				if(isRing(currentWay))
				{
					result.rings.add(currentWay);
					currentWay = new ArrayList<>();
				}
			}
		}
		if(!currentWay.isEmpty())
		{
			result.rest.add(currentWay);
		}

		return result;
	}

	private static boolean isRing(List<?> way)
	{
		return way.get(0).equals(way.get(way.size() - 1));
	}

	/** add <tt>way</tt> to the end of <tt>polyWay</tt>, if necessary in reverse */
	private static void addTo(List<LatLon> way, List<LatLon> polyWay)
	{
		if(polyWay.isEmpty())
		{
			polyWay.addAll(way);
		}
		else
		{
			LatLon addLast = way.get(way.size() - 1);
			LatLon toLast = polyWay.get(polyWay.size() - 1);
			if(addLast.equals(toLast))
			{
				// -1 to not add the last vertex because it has already been added
				ListIterator<LatLon> it = way.listIterator(way.size()-1);
				while(it.hasPrevious())
				{
					polyWay.add(it.previous());
				}
			}
			else
			{
				// +1 to not add the first vertex because it has already been added
				ListIterator<LatLon> it = way.listIterator(+1);
				while(it.hasNext())
				{
					polyWay.add(it.next());
				}
			}
		}
	}

}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/download/ElementGeometryCreator.kt
package de.westnordost.streetcomplete.data.osm.download

import de.westnordost.osmapi.map.data.*
import de.westnordost.streetcomplete.data.meta.OsmAreas
import de.westnordost.streetcomplete.data.osm.ElementGeometry
import de.westnordost.streetcomplete.data.osm.ElementPointGeometry
import de.westnordost.streetcomplete.data.osm.ElementPolygonsGeometry
import de.westnordost.streetcomplete.data.osm.ElementPolylinesGeometry
import de.westnordost.streetcomplete.util.SphericalEarthMath.*
import kotlin.collections.ArrayList

class ElementGeometryCreator(private val wayGeometrySource: WayGeometrySource) {

    fun create(element: Element): ElementGeometry? = when(element) {
        is Node -> create(element)
        is Way -> create(element)
        is Relation -> create(element)
        else -> throw IllegalArgumentException()
    }

    fun create(node: Node) = ElementPointGeometry(node.position)

    fun create(way: Way): ElementGeometry? {
        val nodePositions = wayGeometrySource.getNodePositions(way.id) ?: return null
        val polyline = ArrayList(nodePositions)
        polyline.eliminateDuplicates()
        if (polyline.size < 2) return null

        return if (OsmAreas.isArea(way)) {
            /* ElementGeometry considers polygons that are defined clockwise holes, so ensure that
               it is defined CCW here. */
            if (isRingDefinedClockwise(polyline)) polyline.reverse()
            ElementPolygonsGeometry(arrayListOf(polyline), centerPointOfPolygon(polyline))
        } else {
            ElementPolylinesGeometry(arrayListOf(polyline), centerPointOfPolyline(polyline))
        }
    }

    fun create(relation: Relation): ElementGeometry? {
        return if (OsmAreas.isArea(relation)) {
            createMultipolygonGeometry(relation)
        } else {
            createPolylinesGeometry(relation)
        }
    }

    private fun createMultipolygonGeometry(relation: Relation): ElementPolygonsGeometry? {
        val outer = createNormalizedRingGeometry(relation, "outer", false)
        val inner = createNormalizedRingGeometry(relation, "inner", true)
        if (outer.isEmpty()) return null

        val rings = ArrayList<ArrayList<LatLon>>()
        rings.addAll(outer)
        rings.addAll(inner)

        /* only use first ring that is not a hole if there are multiple
           this is the same behavior as Leaflet or Tangram */
        return ElementPolygonsGeometry(rings, centerPointOfPolygon(outer.first()))
    }

    private fun createPolylinesGeometry(relation: Relation): ElementPolylinesGeometry? {
        val waysNodePositions = getRelationMemberWaysNodePositions(relation)
        val joined = waysNodePositions.joined()

        val polylines = joined.ways
        polylines.addAll(joined.rings)
        if (polylines.isEmpty()) return null

        /* if there are more than one polylines, these polylines are not connected to each other,
           so there is no way to find a reasonable "center point". In most cases however, there
           is only one polyline, so let's just take the first one...
           This is the same behavior as Leaflet or Tangram */
        return ElementPolylinesGeometry(polylines, centerPointOfPolyline(polylines.first()))
    }

    private fun createNormalizedRingGeometry(relation: Relation, role: String, clockwise: Boolean): ArrayList<ArrayList<LatLon>> {
        val waysNodePositions = getRelationMemberWaysNodePositions(relation, role)
        val ringGeometry = waysNodePositions.joined().rings
        ringGeometry.setOrientation(clockwise)
        return ringGeometry
    }

    private fun getRelationMemberWaysNodePositions(relation: Relation): List<List<LatLon>> {
        return relation.members.filter { it.type == Element.Type.WAY }.mapNotNull {
            getValidNodePositions(it.ref)
        }
    }

    private fun getRelationMemberWaysNodePositions(relation: Relation, withRole: String): List<List<LatLon>> {
        return relation.members.filter { it.type == Element.Type.WAY && it.role == withRole }.mapNotNull {
            getValidNodePositions(it.ref)
        }
    }

    private fun getValidNodePositions(wayId: Long): List<LatLon>? {
        val nodePositions = wayGeometrySource.getNodePositions(wayId) ?: return null
        nodePositions.eliminateDuplicates()
        return if (nodePositions.size >= 2) nodePositions else null
    }
}

/** Ensures that all given rings are defined in clockwise/counter-clockwise direction  */
private fun List<MutableList<LatLon>>.setOrientation(clockwise: Boolean) {
    for (ring in this) {
        if (isRingDefinedClockwise(ring) != clockwise) {
            ring.reverse()
        }
    }
}

private fun List<LatLon>.isRing() = first() == last()

private class ConnectedWays(val rings: ArrayList<ArrayList<LatLon>>, val ways: ArrayList<ArrayList<LatLon>>)

/** Returns a list of polylines joined together at their endpoints into rings and ways */
private fun List<List<LatLon>>.joined(): ConnectedWays {
    val nodeWayMap = NodeWayMap(this)

    val rings: ArrayList<ArrayList<LatLon>> = ArrayList()
    val ways: ArrayList<ArrayList<LatLon>> = ArrayList()

    var currentWay: ArrayList<LatLon> = ArrayList()

    while (nodeWayMap.hasNextNode()) {
        val node: LatLon = if (currentWay.isEmpty()) nodeWayMap.getNextNode() else currentWay.last()

        val waysAtNode = nodeWayMap.getWaysAtNode(node)
        if (waysAtNode == null) {
            ways.add(currentWay)
            currentWay = ArrayList()
        } else {
            val way = waysAtNode.first()

            currentWay.join(way)
            nodeWayMap.removeWay(way)

            // finish ring and start new one
            if (currentWay.isRing()) {
                rings.add(currentWay)
                currentWay = ArrayList()
            }
        }
    }
    if (currentWay.isNotEmpty()) {
        ways.add(currentWay)
    }

    return ConnectedWays(rings, ways)
}

/** Join the given adjacent polyline into this polyline */
private fun MutableList<LatLon>.join(way: List<LatLon>) {
    if (isEmpty()) {
        addAll(way)
    } else {
        when {
            last() == way.last() -> addAll(way.asReversed().subList(1, way.size))
            last() == way.first() -> addAll(way.subList(1, way.size))
            first() == way.last() -> addAll(0, way.subList(0, way.size - 1))
            first() == way.first() -> addAll(0, way.asReversed().subList(0, way.size - 1))
            else -> throw IllegalArgumentException("The ways are not adjacent")
        }
    }
}

private fun MutableList<LatLon>.eliminateDuplicates() {
    val it = iterator()
    var prev: LatLon? = null
    while (it.hasNext()) {
        val line = it.next()
        if (prev == null || line.latitude != prev.latitude || line.longitude != prev.longitude) {
            prev = line
        } else {
            it.remove()
        }
    }
}
